### PR TITLE
Fix `Interrupted` falling out of thread crash

### PR DESCRIPTION
**Pull request by roberth**
# Motivation

- Fix a problem with #11618
- Add implicit `checkInterrupt` calls to these loops to pick up on shutdowns sooner

# Context

What I missed is that threads will crash the process if they end in an uncaught exception.
Thank you @lf- for pinging me about this bug!

This does not need a backport unless 2.25 is released first somehow.

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: src/libstore/unix/build/local-derivation-goal.cc [hunk 1]
```diff
@@ -1504,7 +1505,7 @@ void LocalDerivationGoal::startDaemon()
 
     daemonThread = std::thread([this, store]() {
 
-        while (true) {
+        loopUntilInterrupted([this, store]() {
```
Comment (edolstra) [inline src/libstore/unix/build/local-derivation-goal.cc:1508]: It's not really clear to me why we don't just call `checkInterrupt()` in the while loop.

The issue of having an uncaught exception is orthogonal to the interruption issue. It's better to have an appropriate `try { ... } catch (...) { ... }` at top-level in the thread.

File: src/libutil/thread-pool.cc [hunk 2]
```diff
@@ -150,7 +150,10 @@ void ThreadPool::doWork(bool mainThread)
         }
 
         didWork = true;
-    }
+        return Continue;
+    });
+    // We've been interrupted or there's no more work to do; either way we
+    // must return now.
```
Comment (edolstra) [inline src/libutil/thread-pool.cc:156]: Doesn't this create the possibility that if there is an interrupt, the `ThreadPool` will return successfully? I.e. `pool.process()` will return without an error, even though not all of the work items have been executed. And then (for instance) the entire Nix command could succeed if the calling thread doesn't check for interrupts.

**Comment by edolstra** [comment]
Hm, it's not clear to me where `ThreadPool` could have an uncaught `Interrupted` exception. It catches all exceptions from the work item (`try { w(); } catch (...) { exc = std::current_exception(); }`) and I don't see where else an `Interrupted` could be thrown.

Same with the `daemonThread`, though that definitely has an uncaught exception issue if `accept()` fails.

**Comment by roberth** [comment]
> Hm, it's not clear to me where `ThreadPool` could have an uncaught `Interrupted` exception. It catches all exceptions from the work item (`try { w(); } catch (...) { exc = std::current_exception(); }`) and I don't see where else an `Interrupted` could be thrown.
> 
> Same with the `daemonThread`, though that definitely has an uncaught exception issue if `accept()` fails.

It would be thrown by `ignoreExceptionExceptInterrupt`. I guess that's not so obvious, having two layers of weirdness in the name. I think we should change `ignoreException` -> `reportException`, `ExceptInterrupt` -> `RethrowInterrupt`.

**Comment by roberth** [comment]
Also the `ExceptInterrupt` or `RethrowInterrupt` part is a bit suggestive of the wrong assumptions, given that it's currently often used in contexts where `Interrupt` isn't caught. That part - while otherwise harmless - is only relevant when catching any of its superclasses or `...`, so realistically: `std::exception` or `nix::BaseError`.

So we might prefer to "inline" its behavior into the `catch` clauses where it matters, and rely on reviews or perhaps linting to keep overly broad `catch` clauses out of the code. Wdyt?

File: src/libstore/unix/build/local-derivation-goal.cc (hunk 1 above)
Comment (roberth) [reply src/libstore/unix/build/local-derivation-goal.cc:1508]: Here's an alternate fix, https://github.com/NixOS/nix/pull/11665/commits/31fb9a47c5c26b03e9a719aceb31f5c9e83ac980

See also my accidental quote reply https://github.com/NixOS/nix/pull/11665#issuecomment-2404853301

**Comment by lf-** [comment]
> So we might prefer to "inline" its behavior into the `catch` clauses where it matters, and rely on reviews or perhaps linting to keep overly broad `catch` clauses out of the code. Wdyt?

Commentary from the Lix side: we have nuked the most egregious catch in nix (which is std::exception falling out of commands not being an immediate crash with core dump) and caused it to call std::terminate which then prints a stack trace and the exception type id in a custom handler. In general these catch clauses are evil and severely impede debugging: if you have an unknown exception in the daemon you are just screwed, because there's a pile of other exceptions thrown in normal program operation so breakpoints on `__cxa_throw` or so hit dozens of false positives.

(granted printing a stack trace still isn't useful because it's not the stack trace of the throw so it doesn't make the one in main much less bad but it does make every other less broad one less evil; see the library cpptrace for one possible solution to that)

You might want to do similarly.

**Comment by roberth** [comment]
I've got rid of the `loopUntilInterrupted` HOF, which I only ended up using in one case, where it wasn't clear that Interrupted could even be raised.

This is a nice and short diff now.

I think @lf- is making a good point, and I'd focus on removing/replacing bad catch clauses in future fixes and cleanups.
I'll follow this PR up with renames that clarify the behavior of these so-called ignore functions, but keep this diff small.

**Review by edolstra — APPROVED** [review]

**Comment by edolstra** [comment]
@roberth Looks like you need to add an `#include "signals.hh"`.

File: src/libutil/thread-pool.cc [hunk 3]
```diff
@@ -110,10 +110,15 @@ void ThreadPool::doWork(bool mainThread)
                            propagate it. */
                         try {
                             std::rethrow_exception(exc);
+                        } catch (const Interrupted &) {
+                            // The interrupted state may be picked up multiple
```
Comment (cole-h) [inline src/libutil/thread-pool.cc:114]: tiniest of tiny nits, feel free to ignore if you'd rather not wait for CI again :laughing: 

```suggestion
                            // The interrupted state may be picked up by multiple
```

**Pull request merged**